File: src/previous/day98/LRU.ts
type Node<T> = {
  value: T
  next?: Node<T>
  prev?: Node<T>
}

export default class LRU<K, V> {
  private length: number
  private capacity: number
  private head?: Node<V>
  private tail?: Node<V>
  private nodes: Map<K, Node<V>>
  private keys: Map<Node<V>, K>

  constructor(capacity: number = 3) {
    this.length = 0
    this.capacity = capacity
    this.head = this.tail = undefined
    this.nodes = new Map()
    this.keys = new Map()
  }

  update(key: K, value: V): void {
    const node = this.nodes.get(key)
    if (!node) {
      const node: Node<V> = { value }
      this.prepend(node)
      this.length++
      this.trim()

      this.nodes.set(key, node)
      this.keys.set(node, key)
    } else {
      node.value = value
      this.detach(node)
      this.prepend(node)
    }
  }

  get(key: K): V | undefined {
    const node = this.nodes.get(key)
    if (!node) return undefined

    this.detach(node)
    this.prepend(node)
    return node.value
  }

  private detach(node: Node<V>): void {
    if (node.next) node.next.prev = node.prev
    if (node.prev) node.prev.next = node.next
    if (node === this.head) this.head.next = this.head.next
    if (node === this.tail) this.tail.next = this.tail.prev
  }
  private prepend(node: Node<V>): void {
    if (!this.head) {
      this.head = this.tail = node
      return
    }

    node.next = this.head
    this.head.prev = node
    this.head = node
  }
  private trim(): void {
    if (this.length <= this.capacity) return

    const tail = this.tail!
    const key = this.keys.get(tail) as K

    this.nodes.delete(key)
    this.keys.delete(tail)

    this.detach(tail)
    this.length--
  }
}

File: go/day02/LRU/LRU.go
package lru

import "errors"

type Node[V comparable] struct {
	next  *Node[V]
	prev  *Node[V]
	value V
}

type LRU[K comparable, V comparable] struct {
	head     *Node[V]
	tail     *Node[V]
	capacity int
	length   int
	nodes    map[K]*Node[V]
	keys     map[*Node[V]]K
}

func MakeLRU[K comparable, V comparable](capacity int) (*LRU[K, V], error) {
	if capacity < 5 || 1000 <= capacity {
		return nil, errors.New("invalid idx, range 5 ~ 1000")
	}

	return &LRU[K, V]{
		head:     nil,
		tail:     nil,
		capacity: capacity,
		length:   0,
		nodes:    make(map[K]*Node[V]),
		keys:     make(map[*Node[V]]K),
	}, nil
}

func (l *LRU[K, V]) Update(key K, value V) error {
	node := l.nodes[key]
	if node == nil {
		node = &Node[V]{
			value: value,
		}
		l.prepend(node)

		l.length++
		l.trim()

		l.nodes[key] = node
		l.keys[node] = key

	} else {
		node.value = value
		l.detach(node)
		l.prepend(node)
	}

	return nil
}

func (l *LRU[K, V]) Get(key K) V {
	node := l.nodes[key]
	if node == nil {
		var z V
		return z
	}

	l.detach(node)
	l.prepend(node)

	return node.value
}

func (l *LRU[K, V]) detach(node *Node[V]) {
	if node == l.head {
		l.head = l.head.next
	}
	if node == l.tail {
		l.tail = l.tail.prev
	}

	if node.next != nil {
		node.next.prev = node.prev
	}

	if node.prev != nil {
		node.prev.next = node.next
	}
}

func (l *LRU[K, V]) prepend(node *Node[V]) {
	if l.head == nil {
		l.head = node
		l.tail = node
		return
	}

	node.next = l.head
	l.head.prev = node
	l.head = node
}

func (l *LRU[K, V]) trim() error {
	if l.length <= l.capacity {
		return nil
	}

	if l.tail == nil {
		return errors.New("tail is nil")
	}

	tail := l.tail
	key := l.keys[tail]

	delete(l.nodes, key)
	delete(l.keys, tail)

	l.detach(tail)
	l.length--

	return nil
}
